internal/client/client.go

```diff
@@ -58,32 +84,89 @@ func (rf RoundTripFunc) RoundTrip(req *http.Request) (*http.Response, error) {
 	return rf(req)
 }
 
-// RateLimitTransport wraps base transport with rate limiting functionality.
+// RetriesTransport wraps base transport with retry functionality.
 //
-// When a 429 status code is returned by the remote server, the
-// "X-RateLimit-Reset" header is used to determine how long the transport will
-// wait until re-issuing the failed request.
-func RateLimitTransport(base http.RoundTripper) http.RoundTripper {
+// This transport will retry in the following circumstances:
+// Total retries is less than the configured amount
+// AND
+// The configuration specifies to retry on the status OR the error.
+func RetriesTransport(base http.RoundTripper, r RetryOptions) http.RoundTripper {
 	if base == nil {
 		base = http.DefaultTransport
 	}
-	return rehttp.NewTransport(base, retry, delay)
+
+	// Configure a retry transport that will retry if:
+	// Total retries is less than the configured amount
+	// AND
+	// The configuration specifies to retry on the status OR the error
+	tr := rehttp.NewTransport(
+		base,
+		rehttp.RetryAll(
+			rehttp.RetryMaxRetries(r.MaxRetries),
+			rehttp.RetryAny(
+				rehttp.RetryStatuses(r.Statuses...),
+				rehttp.RetryIsErr(retryErrors),
+			),
+		),
+		backoffDelay(),
+	)
+
+	return tr
 }
 
-func retry(attempt rehttp.Attempt) bool {
-	if attempt.Response == nil {
-		return false
+// Matches the error returned by net/http when the configured number of
+// is exhausted.
+var redirectsErrorRe = regexp.MustCompile(`stopped after \d+ redirects\z`)
+
+// Matches the error returned by net/http when the TLS certificate is not trusted.
+// When version 1.20 is the minimum supported this can be replaced by tls.CertificateVerificationError.
```

Comment:
Note that if you're making a new major version it is acceptable to have it support only Go 1.20+. Specially since Go 1.21 is expected for August 2023, a mere 3 months away, which will drop support for 1.19.

Reply:
Yeah I think that's a fair shout, I'll maybe circle back to dropping 1.19 after checking one some of the users of this package have their compile target.